pkg/noun/v3/manage.c

```diff
@@ -0,0 +1,111 @@
+/// @file
+
+#include "pkg/noun/v3/manage.h"
+
+#include "pkg/noun/v2/jets.h"
+#include "pkg/noun/v2/nock.h"
+#include "pkg/noun/v2/vortex.h"
+
+#include "pkg/noun/v3/allocate.h"
+#include "pkg/noun/v3/hashtable.h"
+#include "pkg/noun/version.h"
+#include "pkg/noun/v3/vortex.h"
+#include <v2/allocate.h>
+
+/* u3m_v3_migrate: perform loom migration if necessary.
+*/
+void
+u3m_v3_migrate()
+{
+  fprintf(stderr, "loom: memoization migration running...\r\n");
+
+
+  c3_w ver_w = *(u3_Loom + u3C.wor_i - 1);
+  c3_w *mem_w = u3_Loom + u3a_v3_walign;
+  c3_w  len_w = u3C.wor_i - u3a_v3_walign;
+  c3_w  suz_w = c3_wiseof(u3v_v2_home);
+  c3_w *mut_w = c3_align(mem_w + len_w - suz_w, u3a_v3_balign, C3_ALGLO);
+
+  //  old road
+  u3v_v2_home* hum_u = (u3v_v2_home*)mut_w;
+  u3a_v2_road* rud_u = &hum_u->rod_u;
+  size_t ruz_t = sizeof(u3a_v2_road);
+
+  //  set v2 globals
+  u3H_v2 = (void *)mut_w;
```

Comment:
This is kind of a hack because `u3[H,R]_v2` don't contain all the members in the current home/road struct versions (in the main `pkg/noun` folder), and yet we're casting them to the current struct versions anyways and using them to set the global pointers. The reason for this is that the `u3[j,n]_v2_reclaim` functions which we call on lines 42 and 43 are actually aliases to `u3[j,n]_reclaim` (the current versions of the functions). The current functions require the global `u3H` and `u3R` pointers to be set. It just _so happens_ that these functions don't touch any of the struct members which were added to the `v3` (i.e., latest) versions of the home and road structs, so effectively everything just works. If we wanted to avoid this sort of "lying" we'd have to make `v2` versions of all the functions that rely on those global pointers instead of using aliases.

Reply:
Setting v2 here is not a hack, we're initializing a v2 loom. The hack is below, setting `u3R` to alias `u3R_v2`. The reason we're doing that is that `u3{j,n}_v2_reclaim` are aliases, and need a current road pointer to function. It might be better to implement the v2 reclaim functions as wrappers around v3 functions that first set `u3R` -- that encodes the hack more precisely.